File: 0017-Letter Combinations of a Phone Number/0017-Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> result;
        if (digits.size() == 0) {
            return result;
        }
        dfs(digits, 0, "", result);
        return result;
    }
    
private:
    vector<string> table = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    void dfs(string digits, int start, string path, vector<string>& result) {
        if (start == digits.size()) {
            result.push_back(path);
            return;
        }
        
        int index = digits[start] - '0';
        for (int i = 0; i < table[index].size(); ++i) {
            dfs(digits, start + 1, path + table[index][i], result);
        }
    }    
};

File: 0001-0100/0017-Letter Combinations of a Phone Number/0017-Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        result = []
        table = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        def dfs(start, path):
            if start == len(digits):
                result.append(path)
                return
            
            for c in table[int(digits[start])]:
                dfs(start + 1, path + c)
        if digits:
            dfs(0, '')
        return result
